aeson-1.0.0.0 support

There are two compilation error against the newest aeson, they can be fixed in different ways.

* Add `{-# OVERLAPS #-}` pragmas for newer GHCs and enable `OverlappingInstances` for older ones.
* Conditionally replace the instances with JSONKey instances to use aeson's built in instances for maps

Which one should we go for?

```
/Users/adam/repos/3rd/snap/src/Snap/Snaplet/Auth/Backends/JsonFile.hs:303:7: error:
    • Overlapping instances for ToJSON UserIdCache
        arising from a use of ‘.=’
      Matching instances:
        instance (ToJSON v, ToJSONKey k) => ToJSON (Map k v)
          -- Defined in ‘aeson-1.0.0.0:Data.Aeson.Types.ToJSON’
        instance ToJSON UserIdCache
          -- Defined at src/Snap/Snaplet/Auth/Backends/JsonFile.hs:92:10
    • In the expression: "uidCache" .= uidCache uc
      In the first argument of ‘object’, namely
        ‘["uidCache" .= uidCache uc, "loginCache" .= loginCache uc,
          "tokenCache" .= tokenCache uc, "uidCounter" .= uidCounter uc]’
      In the expression:
        object
          ["uidCache" .= uidCache uc, "loginCache" .= loginCache uc,
           "tokenCache" .= tokenCache uc, "uidCounter" .= uidCounter uc]

/Users/adam/repos/3rd/snap/src/Snap/Snaplet/Auth/Backends/JsonFile.hs:314:11: error:
    • Overlapping instances for FromJSON UserIdCache
        arising from a use of ‘.:’
      Matching instances:
        instance (FromJSONKey k, Ord k, FromJSON v) => FromJSON (Map k v)
          -- Defined in ‘aeson-1.0.0.0:Data.Aeson.Types.FromJSON’
        instance FromJSON UserIdCache
          -- Defined at src/Snap/Snaplet/Auth/Backends/JsonFile.hs:95:10
    • In the second argument of ‘(<$>)’, namely ‘v .: "uidCache"’
      In the first argument of ‘(<*>)’, namely
        ‘UserCache <$> v .: "uidCache"’
      In the first argument of ‘(<*>)’, namely
        ‘UserCache <$> v .: "uidCache" <*> v .: "loginCache"’
```

--- a/snap.cabal
+++ b/snap.cabal
@@ -103,7 +103,7 @@ Library
     Snap.Snaplet.Session.SecureCookie
 
   build-depends:
-    aeson                     >= 0.6      && < 0.12,
+    aeson                     >= 0.6      && < 1.1,
     attoparsec                >= 0.10     && < 0.14,
     base                      >= 4        && < 5,
     bytestring                >= 0.9.1    && < 0.11,

--- a/src/Snap/Snaplet/Auth/Backends/JsonFile.hs
+++ b/src/Snap/Snaplet/Auth/Backends/JsonFile.hs
@@ -89,11 +89,15 @@ mkJsonAuthMgr fp = do
 ------------------------------------------------------------------------------
 type UserIdCache = Map UserId AuthUser
 
+#if !MIN_VERSION_aeson(1,0,0)
+-- In aeson >= 1 these instances are not needed because we have
+-- derived ToJSONKey/FromJSONKey instances for UserId.
 instance ToJSON UserIdCache where
   toJSON m = toJSON $ HM.toList m
 
 instance FromJSON UserIdCache where
   parseJSON = fmap HM.fromList . parseJSON
+#endif
 
 ------------------------------------------------------------------------------
 type LoginUserCache = Map Text UserId

--- a/src/Snap/Snaplet/Auth/Types.hs
+++ b/src/Snap/Snaplet/Auth/Types.hs
@@ -3,6 +3,7 @@
 {-# LANGUAGE ExistentialQuantification  #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE StandaloneDeriving         #-}
 
 module Snap.Snaplet.Auth.Types where
 
@@ -113,6 +114,10 @@ instance Show AuthFailure where
 newtype UserId = UserId { unUid :: Text }
   deriving ( Read, Show, Ord, Eq, FromJSON, ToJSON, Hashable )
 
+#if MIN_VERSION_aeson(1,0,0)
+deriving instance FromJSONKey UserId
+deriving instance ToJSONKey UserId
+#endif
 
 ------------------------------------------------------------------------------
 -- | This will be replaced by a role-based permission system.

